Epic 2.9: Improved Syllabus Generator with Context Chaining (Prototype)

sairishi-exe opened the pull request:
## Description

Implements a new syllabus generator tool using a hybrid architecture that combines context chaining, prompt enhancement, and optional field handling. The tool uses a three-chain LLM pipeline with shared context or a "source of truth" to generate comprehensive and coherent syllabi with structured outputs.

## Related Issue

EPIC 2.9 

## Type of Change

- [x] **New feature**: A non-breaking change that adds functionality.
- [ ] **Documentation update**: Changes or updates to documentation.

## Proposed Solution

Implemented a hybrid architecture for syllabus generation:
- Context chaining between three LLM chains for progressive content refinement
- Prompt enhancement using dynamic templates and fallback defaults
- Optional field handling with smart defaults and validation

**The transition from a fully parallel architecture to a hybrid one, where context 
is passed using the following approach:**

1. Chain 1: Initial course content and policy generation (`course_content` is used for shared context)
2. Chain 2: Content-dependent components with enhanced context
3. Chain 3: Schedule generation with activity alignment

## How to Test

Manual endpoint testing can be performed using the Request Schema for Syllabus Generator
in [app/tools/README.md](https://github.com/marvelai-org/marvel-ai-backend/blob/Develop/app/tools/README.md)

## Functionality Showcase

Here is a Loom Video showcasing the usage of the Syllabus Generator Tool:

https://www.loom.com/share/e27b1cab41a14e58ba7b9d146de845d6

## Unit Tests

No automated unit tests or integration tests have been implemented yet. Current testing has been performed manually through endpoint testing.

Future test coverage should include:
- Input validation tests
- Chain integration tests
- Error handling scenarios
- Multi-language support verification
- Optional field handling tests

## Documentation Updates

- [x] Yes
- [ ] No

The [README.md](https://github.com/marvelai-org/marvel-ai-backend/blob/Develop/app/tools/README.md) file in the app/tools folder needs to be updated in the output schema.

## Checklist

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ]  I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

## Additional Information
**Check out the proposal document below for more information:**
https://www.notion.so/Task-3-Diagnose-prototype-and-propose-1a2456247e0a8095b616d80590251e9e

Commit:
User checkpoint: Configure Replit environment for uvicorn app deployment.

Commit:
User checkpoint: Update Replit configuration to run uvicorn app.main:app.

stevenrayhinojosa-gmail-com committed:
Assistant checkpoint: Add Common Core standards to quiz questions

Assistant generated file changes:
- app/tools/multiple_choice_quiz_generator/prompt/multiple_choice_quiz_generator_prompt.txt: Update prompt to include Common Core standards
- app/tools/multiple_choice_quiz_generator/tools.py: Add standards field to QuizQuestion model

---

User prompt:

in the Multiple Choice Quiz Generator I want to add a new feature that adds the common core standards to the quiz question

Commit:
feat: improved syllabus generation with context-chaining

stevenrayhinojosa-gmail-com commented:
Fantastic job, ran well, sending to @yunusj

AaronSosaRamos reviewed (approved):
Very good work!
A showcase for your feature is needed through a Loom Video and with the given Benchmarks. 
Thank you.

sairishi-exe commented:
> Very good work! A showcase for your feature is needed through a Loom Video and with the given Benchmarks. Thank you.
Sure will do that and also add some tests. Thank you!

AaronSosaRamos commented:
Hey @sairishi-exe, I think you tried to upload your video but nothing is presented yet. We are still waiting for your outstanding showcase on this!

sairishi-exe commented:
> Hey @sairishi-exe, I think you tried to upload your video but nothing is presented yet. We are still waiting for your outstanding showcase on this!

Hello @AaronSosaRamos, apologies for the delay; I was busy with Mission Week 2 Image Generation. I have uploaded the Loom Video showcasing functionality.

Here is the full response body from the Loom Video:
```json
{
  "data": {
    "course_information": {
      "course_title": "Linear Regression: From Foundations to Advanced Applications",
      "grade_level": "College",
      "description": "Unlock the power of predictive modeling with this comprehensive linear regression course. Gain a deep understanding of how to build, evaluate, and refine robust linear regression models, progressing from fundamental concepts to advanced techniques like regularization and model validation.  Develop practical skills by applying your knowledge to real-world datasets and effectively communicating your findings.  Ideal for anyone seeking to master a core statistical method crucial for data analysis and informed decision-making."
    },
    "course_description_objectives": {
      "objectives": [
        "Understand the theoretical foundations of linear regression analysis.",
        "Develop proficiency in performing linear regression analysis on datasets using appropriate techniques.",
        "Interpret and communicate statistical results effectively in the context of real-world problems."
      ],
      "intended_learning_outcomes": [
        {
          "category": "Technical Skills",
          "outcomes": [
            "Ability to build and validate linear regression models using least squares and regularization techniques (within 8 weeks).",
            "Proficiency in calculating and interpreting regression coefficients, R-squared, adjusted R-squared, and other relevant statistics (within 8 weeks).",
            "Skill in performing hypothesis testing for regression coefficients and model significance (within 3 weeks).",
            "Ability to apply variable selection techniques in multivariate regression to identify significant predictors (within 4 weeks).",
            "Competence in using diagnostic tools like residual analysis, leverage, and Cook's distance to assess model adequacy (within 5 weeks).",
            "Proficiency in implementing regularization techniques like Ridge, Lasso, and Elastic Net to address overfitting and multicollinearity (within 6 weeks).",
            "Ability to utilize cross-validation and train-test split methods for model validation and selection (within 7 weeks)."
          ]
        },
        {
          "category": "Analytical Skills",
          "outcomes": [
            "Ability to critically evaluate model assumptions and identify potential violations (within 5 weeks).",
            "Skill in interpreting the impact of multiple predictors on the dependent variable in multivariate regression (within 4 weeks).",
            "Ability to analyze residuals to detect heteroscedasticity, non-normality, and other issues (within 5 weeks).",
            "Capacity to discern the effects of multicollinearity on regression results (within 4 weeks).",
            "Skill in selecting the most appropriate model based on validation metrics such as RMSE and MAE (within 7 weeks)."
          ]
        },
        {
          "category": "Practical Skills",
          "outcomes": [
            "Ability to apply linear regression techniques to real-world datasets for predictive modeling and analysis (within 8 weeks).",
            "Skill in performing data cleaning and preprocessing steps necessary for regression analysis (within 8 weeks).",
            "Competence in interpreting and communicating regression results in a clear and concise manner to both technical and non-technical audiences (within 8 weeks)."
          ]
        }
      ]
    },
    "course_content": [
      {
        "unit_time": "week",
        "unit_time_value": 1,
        "main_topic": "Introduction to Linear Regression",
        "subtopics": [
          "Basic concepts of linear regression",
          "Use cases and applications of linear regression",
          "Dependent and independent variables",
          "Scatter plots and visualizing relationships",
          "Correlation coefficient and its interpretation",
          "Introduction to the linear equation (Y = a + bX)",
          "Assumptions of linear regression"
        ]
      },
      {
        "unit_time": "week",
        "unit_time_value": 2,
        "main_topic": "Least Squares Method",
        "subtopics": [
          "Derivation of the least squares estimators",
          "Calculating the slope and intercept",
          "Minimizing the sum of squared errors",
          "Understanding the cost function",
          "Properties of least squares estimators",
          "Introduction to R-squared and its interpretation"
        ]
      },
      {
        "unit_time": "week",
        "unit_time_value": 3,
        "main_topic": "Hypothesis Testing",
        "subtopics": [
          "Hypothesis testing for regression coefficients",
          "t-tests and p-values",
          "Confidence intervals for slope and intercept",
          "Testing the significance of the regression model",
          "F-statistic and its interpretation"
        ]
      },
      {
        "unit_time": "week",
        "unit_time_value": 4,
        "main_topic": "Multivariate Regression",
        "subtopics": [
          "Extending linear regression to multiple predictors",
          "Interpreting coefficients in multivariate regression",
          "Multicollinearity and its effects",
          "Variable selection techniques",
          "Adjusted R-squared"
        ]
      },
      {
        "unit_time": "week",
        "unit_time_value": 5,
        "main_topic": "Diagnostics and Residual Analysis",
        "subtopics": [
          "Examining residuals for model adequacy",
          "Identifying outliers and influential observations",
          "Leverage and Cook's distance",
          "Heteroscedasticity and its remedies",
          "Normality assumptions and tests"
        ]
      },
      {
        "unit_time": "week",
        "unit_time_value": 6,
        "main_topic": "Regularization Techniques",
        "subtopics": [
          "Introduction to regularization",
          "Ridge regression",
          "Lasso regression",
          "Elastic net regression",
          "Bias-variance trade-off"
        ]
      },
      {
        "unit_time": "week",
        "unit_time_value": 7,
        "main_topic": "Model Validation",
        "subtopics": [
          "Cross-validation techniques",
          "Train-test split",
          "Model performance metrics (RMSE, MAE, etc.)",
          "Selecting the best model"
        ]
      },
      {
        "unit_time": "week",
        "unit_time_value": 8,
        "main_topic": "Capstone Project",
        "subtopics": [
          "Applying linear regression to a real-world dataset",
          "Data cleaning and preprocessing",
          "Model building and evaluation",
          "Interpretation and communication of results",
          "Project presentation"
        ]
      }
    ],
    "policies_procedures": {
      "attendance_policy": "While attendance is not formally graded, active participation in discussions is a crucial component of the learning process and is factored into the overall course grade.  Regular attendance is strongly encouraged to maximize learning and engagement with the material.  If you anticipate missing a class, please inform the instructor in advance.  Any missed material is the student's responsibility to obtain and review.",
      "late_submission_policy": "All assignments are due at the specified date and time. Late submissions will be penalized [Specify penalty - e.g., 10% deduction per day late, up to a maximum of 50%].  In exceptional circumstances, such as documented illness or family emergencies, please contact the instructor before the deadline to discuss possible extensions. Extensions will be granted on a case-by-case basis.",
      "academic_honesty": "All work submitted for this course must be your own original work. Collaboration on assignments is permitted only when explicitly authorized by the instructor.  Any instances of plagiarism, cheating, or other forms of academic dishonesty will not be tolerated and will result in [Specify consequences - e.g., a failing grade on the assignment, a failing grade in the course, referral to the university's academic integrity board].  Students are expected to familiarize themselves with the university's academic integrity policies."
    },
    "assessment_grading_criteria": {
      "assessment_methods": [
        {
          "type_assessment": "Midterm Exam (Weeks 1-4)",
          "weight": 30
        },
        {
          "type_assessment": "Final Exam (Weeks 5-8)",
          "weight": 30
        },
        {
          "type_assessment": "Homework Assignments (Weekly)",
          "weight": 10
        },
        {
          "type_assessment": "Capstone Project (Week 8)",
          "weight": 30
        }
      ],
      "grading_scale": {
        "Midterm Exam (Weeks 1-4)": {
          "Content": "Covers topics from weeks 1-4: Introduction to Linear Regression, Least Squares Method, Hypothesis Testing, and Multivariate Regression.",
          "Format": "Combination of multiple-choice questions, short answer questions requiring calculations and interpretations, and problem-solving related to applying the concepts learned.",
          "Grading": "Points assigned to each question based on difficulty. Total score contributes 30% to the final grade."
        },
        "Final Exam (Weeks 5-8)": {
          "Content": "Covers topics from weeks 5-8: Diagnostics and Residual Analysis, Regularization Techniques, Model Validation, and conceptual overview connecting to earlier topics.  Includes interpretation of outputs and application of techniques.",
          "Format": "Similar to Midterm, with more emphasis on application, analysis, and interpretation of results.  May include case-study based questions related to model building and validation.",
          "Grading": "Points assigned to each question based on difficulty. Total score contributes 30% to the final grade."
        },
        "Homework Assignments (Weekly)": {
          "Content": "Weekly assignments covering practical application and problem-solving related to the week's topic. Will involve calculations, interpretations, and potentially using software for basic linear regression analysis.",
          "Format": "Problem sets with clearly defined instructions and expected outputs. May involve coding exercises for data analysis and visualization.",
          "Grading": "Each assignment graded based on correctness and completeness. Average score across all assignments contributes 10% to the final grade."
        },
        "Capstone Project (Week 8)": {
          "Content": "A comprehensive project requiring students to apply linear regression to a real-world dataset. Involves data cleaning, preprocessing, model building, evaluation, interpretation, and presentation of findings.",
          "Format": "Students will work individually on a chosen dataset.  Deliverables include a written report detailing the process and findings, along with code used for analysis and a short presentation summarizing the project.",
          "Grading": "Rubric-based assessment considering data preprocessing, model selection, analysis, interpretation of results, clarity of the report, and presentation quality. Project contributes 30% to the final grade."
        }
      }
    },
    "learning_resources": [
      {
        "title": "Introduction to Linear Regression Analysis",
        "author": "Douglas C. Montgomery, Elizabeth A. Peck, G. Geoffrey Vining",
        "year": 2021
      },
      {
        "title": "Applied Linear Regression Models",
        "author": "Michael H. Kutner, Christopher J. Nachtsheim, John Neter, William Li",
        "year": 2004
      },
      {
        "title": "Linear Regression with R",
        "author": "Julian J. Faraway",
        "year": 2015
      },
      {
        "title": "Discovering Statistics Using R",
        "author": "Andy Field, Jeremy Miles, Zoe Field",
        "year": 2012
      },
      {
        "title": "An Introduction to Statistical Learning",
        "author": "Gareth James, Daniela Witten, Trevor Hastie, Robert Tibshirani",
        "year": 2013
      },
      {
        "title": "The Elements of Statistical Learning",
        "author": "Trevor Hastie, Robert Tibshirani, Jerome Friedman",
        "year": 2009
      }
    ],
    "course_schedule": [
      {
        "unit_time": "week",
        "unit_time_value": 1,
        "main_topic": "Introduction to Linear Regression",
        "topics": [
          "Basic concepts of linear regression",
          "Use cases and applications of linear regression",
          "Dependent and independent variables",
          "Scatter plots and visualizing relationships",
          "Correlation coefficient and its interpretation",
          "Introduction to the linear equation (Y = a + bX)",
          "Assumptions of linear regression"
        ],
        "activities": [
          "Introductory lecture",
          "Hands-on exercises",
          "Homework 1 assigned"
        ],
        "objectives": [
          "Understand the theoretical foundations of linear regression analysis."
        ],
        "assessments": [
          "Homework 1 assigned"
        ]
      },
      {
        "unit_time": "week",
        "unit_time_value": 2,
        "main_topic": "Least Squares Method",
        "topics": [
          "Derivation of the least squares estimators",
          "Calculating the slope and intercept",
          "Minimizing the sum of squared errors",
          "Understanding the cost function",
          "Properties of least squares estimators",
          "Introduction to R-squared and its interpretation"
        ],
        "activities": [
          "Lecture on Least Squares Method",
          "Practical exercises on calculating slope and intercept",
          "Homework 2 assigned"
        ],
        "objectives": [
          "Develop proficiency in performing linear regression analysis on datasets using appropriate techniques."
        ],
        "assessments": [
          "Homework 2 assigned"
        ]
      },
      {
        "unit_time": "week",
        "unit_time_value": 3,
        "main_topic": "Hypothesis Testing",
        "topics": [
          "Hypothesis testing for regression coefficients",
          "t-tests and p-values",
          "Confidence intervals for slope and intercept",
          "Testing the significance of the regression model",
          "F-statistic and its interpretation"
        ],
        "activities": [
          "Lecture on Hypothesis Testing",
          "Exercises on t-tests and p-values",
          "Homework 3 assigned"
        ],
        "objectives": [
          "Skill in performing hypothesis testing for regression coefficients and model significance (within 3 weeks)."
        ],
        "assessments": [
          "Homework 3 assigned"
        ]
      },
      {
        "unit_time": "week",
        "unit_time_value": 4,
        "main_topic": "Multivariate Regression",
        "topics": [
          "Extending linear regression to multiple predictors",
          "Interpreting coefficients in multivariate regression",
          "Multicollinearity and its effects",
          "Variable selection techniques",
          "Adjusted R-squared"
        ],
        "activities": [
          "Lecture on Multivariate Regression",
          "Case study examples",
          "Homework 4 assigned",
          "Midterm Exam (Weeks 1-4)"
        ],
        "objectives": [
          "Ability to apply variable selection techniques in multivariate regression to identify significant predictors (within 4 weeks).",
          "Skill in interpreting the impact of multiple predictors on the dependent variable in multivariate regression (within 4 weeks).",
          "Capacity to discern the effects of multicollinearity on regression results (within 4 weeks)."
        ],
        "assessments": [
          "Homework 4 assigned",
          "Midterm Exam (Weeks 1-4)"
        ]
      },
      {
        "unit_time": "week",
        "unit_time_value": 5,
        "main_topic": "Diagnostics and Residual Analysis",
        "topics": [
          "Examining residuals for model adequacy",
          "Identifying outliers and influential observations",
          "Leverage and Cook's distance",
          "Heteroscedasticity and its remedies",
          "Normality assumptions and tests"
        ],
        "activities": [
          "Lecture on Diagnostics and Residual Analysis",
          "Practical exercises on residual analysis",
          "Homework 5 assigned"
        ],
        "objectives": [
          "Competence in using diagnostic tools like residual analysis, leverage, and Cook's distance to assess model adequacy (within 5 weeks).",
          "Ability to critically evaluate model assumptions and identify potential violations (within 5 weeks).",
          "Ability to analyze residuals to detect heteroscedasticity, non-normality, and other issues (within 5 weeks)."
        ],
        "assessments": [
          "Homework 5 assigned"
        ]
      },
      {
        "unit_time": "week",
        "unit_time_value": 6,
        "main_topic": "Regularization Techniques",
        "topics": [
          "Introduction to regularization",
          "Ridge regression",
          "Lasso regression",
          "Elastic net regression",
          "Bias-variance trade-off"
        ],
        "activities": [
          "Lecture on Regularization Techniques",
          "Hands-on exercises with Ridge, Lasso, and Elastic Net",
          "Homework 6 assigned"
        ],
        "objectives": [
          "Proficiency in implementing regularization techniques like Ridge, Lasso, and Elastic Net to address overfitting and multicollinearity (within 6 weeks)."
        ],
        "assessments": [
          "Homework 6 assigned"
        ]
      },
      {
        "unit_time": "week",
        "unit_time_value": 7,
        "main_topic": "Model Validation",
        "topics": [
          "Cross-validation techniques",
          "Train-test split",
          "Model performance metrics (RMSE, MAE, etc.)",
          "Selecting the best model"
        ],
        "activities": [
          "Lecture on Model Validation",
          "Practical exercises on cross-validation and train-test split",
          "Homework 7 assigned"
        ],
        "objectives": [
          "Ability to utilize cross-validation and train-test split methods for model validation and selection (within 7 weeks).",
          "Skill in selecting the most appropriate model based on validation metrics such as RMSE and MAE (within 7 weeks)."
        ],
        "assessments": [
          "Homework 7 assigned"
        ]
      },
      {
        "unit_time": "week",
        "unit_time_value": 8,
        "main_topic": "Capstone Project",
        "topics": [
          "Applying linear regression to a real-world dataset",
          "Data cleaning and preprocessing",
          "Model building and evaluation",
          "Interpretation and communication of results",
          "Project presentation"
        ],
        "activities": [
          "Work on Capstone Project",
          "Project presentations",
          "Final Exam (Weeks 5-8)"
        ],
        "objectives": [
          "Ability to build and validate linear regression models using least squares and regularization techniques (within 8 weeks).",
          "Proficiency in calculating and interpreting regression coefficients, R-squared, adjusted R-squared, and other relevant statistics (within 8 weeks).",
          "Ability to apply linear regression techniques to real-world datasets for predictive modeling and analysis (within 8 weeks).",
          "Skill in performing data cleaning and preprocessing steps necessary for regression analysis (within 8 weeks).",
          "Competence in interpreting and communicating regression results in a clear and concise manner to both technical and non-technical audiences (within 8 weeks)."
        ],
        "assessments": [
          "Capstone Project due",
          "Final Exam (Weeks 5-8)"
        ]
      }
    ]
  }
}
```

Commit:
refactor: changed the pydantic model syllabus schema to be more flexible

Commit:
fix: made fallback functions more stable

Commit:
chore: changed .gitignore to exclude any credential files
